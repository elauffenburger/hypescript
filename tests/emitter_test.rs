use hypescript::{emitter, parser};

#[macro_use]
mod macros {
    macro_rules! insta_test {
        ($code:expr) => {{
            let parsed = parser::parse($code).unwrap();

            insta::assert_debug_snapshot!(emitter::Emitter::new().emit(parsed).unwrap());
        }};
    }
}

#[test]
fn can_emit_src() {
    insta_test!(
        r#"
        function main() {
            console.log("hello, world!");
        } 

        main();
    "#
    )
}

#[test]
fn can_emit_complex_src() {
    insta_test!(
        r#"
        interface Foo {
            str: string;
            num: number;
            bar: Bar;
            baz: Baz;
            qux?: string | number & Baz;
        }

        interface Bar {
            str: string;
        }

        interface Baz {
            name: string;
        }

        function foo(a: string, b: number): number {
            let ay = 5;
            let bee = "bar";

            return ay;
        }

        function blah() {
            let foo = "asdf";
            let bar = foo;
            bar = "bar";

            return foo;
        }

        function blah2() {}

        function blah3() {
            let foo: number = 5;

            return foo;
        }

        function returnsFn() {
            return function() {
                console.log("in nested!");
            };
        }

        function run(): void {
            function baz() {
                console.log("in baz!");
            }

            let obj = { 
                foo: "bar", 
                baz: 5, 
                qux: { 
                    a: "a",
                    foo: function() {
                        console.log("in foo!");
                    }
                } 
            };

            obj.qux.a = "hello, world!";

            console.log(obj.qux.a);
            obj.qux.foo();

            baz();

            returnsFn()();

            let foo: Foo = {
                str: "str",
                num: 5,
                bar: {
                    str: "hello",
                },
                baz: {
                    name: "world!",
                },
            };

            console.log(foo.bar.str);
            console.log(foo.baz.name);

            let bar = {
                name: "foo",
                sayName: function() {
                    console.log(this.getName());
                },
                getName: function(): string {
                    return this.name;
                },
            };

            bar.sayName();
        }

        run();
    "#
    )
}

#[test]
fn can_use_for_loop() {
    insta_test!(
        r#"
        for (let i = 0; i < 10; i++) {
            console.log(i);
        }
        "#
    )
}

#[test]
fn can_emit_fizzbuzz() {
    insta_test!(
        r#"
            function fizzbuzz(n: number): void {
                for (let i = 1; i < n + 1; i++) {
                    let fizz = i % 3 == 0;
                    let buzz = i % 5 == 0;

                    if (fizz && buzz) {
                        console.log("fizzbuzz");
                    } else if (fizz) {
                        console.log("fizz");
                    } else if (buzz) {
                        console.log("buzz");
                    } else {
                        console.log(i);
                    }
                }
            }
        "#
    )
}

#[test]
fn can_iife() {
    insta_test!(
        r#"
            (function(){ return 42; })()
        "#
    )
}

#[test]
fn can_emit_expr_with_ops() {
    insta_test!(
        r#"
            let foo = {
                bar: {
                    str: "hello, world!"
                },
                baz: {
                    name: "i'm a baz!"
                }
            }

            console.log(foo.bar.str);
            console.log(foo.baz.name);
        "#
    )
}

#[test]
fn can_emit_expr_with_assignment() {
    insta_test!(
        r#"
            let foo = {
                name: "i'm a bar!"
            };

            foo.name = "i'm a foo!";
        "#
    )
}

#[test]
fn can_emit_subexpr() {
    insta_test!(
        r#"
            (42);
        "#
    )
}
