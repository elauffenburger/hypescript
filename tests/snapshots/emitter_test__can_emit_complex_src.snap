---
source: tests/emitter_test.rs
expression: "emitter::Emitter::new().emit(parsed).unwrap()"
---
EmitterResult {
    code: "#include <stdlib.h>\n#include <stdio.h>\n#include <string>\n#include <vector>\n#include <algorithm>\n#include <memory>\n#include \"runtime.hpp\"\nint main() {\nTsFunction* _foo = new TsFunction(\"foo\",TsCoreHelpers::toVector<TsFunctionParam>({TsFunctionParam(\"a\", 0)TsFunctionParam(\"b\", 0)}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nauto _ay = new TsNum(5);\nauto _bee = new TsString(\"bar\");\nreturn _ay\n;\n});\nTsFunction* _blah = new TsFunction(\"blah\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nauto _foo = new TsString(\"asdf\");\nauto _bar = _foo;\n_bar = new TsString(\"bar\");\nreturn _foo\n;\n});\nTsFunction* _blah2 = new TsFunction(\"blah2\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nreturn NULL;\n});\nTsFunction* _blah3 = new TsFunction(\"blah3\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nauto _foo = new TsNum(5);\nreturn _foo\n;\n});\nTsFunction* _returnsFn = new TsFunction(\"returnsFn\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nreturn new TsFunction(\"\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nconsole->getFieldValue(\"log\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({TsFunctionArg(\"sdjkl\", new TsString(\"in nested!\"))}));\nreturn NULL;\n})\n;\n});\nTsFunction* _run = new TsFunction(\"run\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nTsFunction* _baz = new TsFunction(\"baz\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nconsole->getFieldValue(\"log\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({TsFunctionArg(\"sdjkl\", new TsString(\"in baz!\"))}));\nreturn NULL;\n});\nauto _obj = new TsObject(1, TsCoreHelpers::toVector<ToObjectField*>({new TsObjectField(TsObjectFieldDescriptor(TsString(\"foo\"), 0)new TsString(\"bar\")), new TsObjectField(TsObjectFieldDescriptor(TsString(\"baz\"), 0)new TsNum(5)), new TsObjectField(TsObjectFieldDescriptor(TsString(\"qux\"), 0)new TsObject(1, TsCoreHelpers::toVector<ToObjectField*>({new TsObjectField(TsObjectFieldDescriptor(TsString(\"a\"), 0)new TsString(\"a\")), new TsObjectField(TsObjectFieldDescriptor(TsString(\"foo\"), 0)new TsFunction(\"\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nconsole->getFieldValue(\"log\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({TsFunctionArg(\"sdjkl\", new TsString(\"in foo!\"))}));\nreturn NULL;\n}))})))}));\n_obj->getFieldValue(\"qux\")->getFieldValue(\"a\")->setFieldValue(\"a\", new TsString(\"hello, world!\"));\nconsole->getFieldValue(\"log\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({TsFunctionArg(\"sdjkl\", _obj->getFieldValue(\"qux\")->getFieldValue(\"a\"))}));\n_obj->getFieldValue(\"qux\")->getFieldValue(\"foo\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({}));\n_baz->invoke(TsCoreHelpers::toVector<TsFunctionArg>({}));\n_returnsFn->invoke(TsCoreHelpers::toVector<TsFunctionArg>({}))->invoke(TsCoreHelpers::toVector<TsFunctionArg>({}));\nauto _foo = new TsObject(1, TsCoreHelpers::toVector<ToObjectField*>({new TsObjectField(TsObjectFieldDescriptor(TsString(\"str\"), 0)new TsString(\"str\")), new TsObjectField(TsObjectFieldDescriptor(TsString(\"num\"), 0)new TsNum(5)), new TsObjectField(TsObjectFieldDescriptor(TsString(\"bar\"), 0)new TsObject(1, TsCoreHelpers::toVector<ToObjectField*>({new TsObjectField(TsObjectFieldDescriptor(TsString(\"str\"), 0)new TsString(\"hello\"))}))), new TsObjectField(TsObjectFieldDescriptor(TsString(\"baz\"), 0)new TsObject(1, TsCoreHelpers::toVector<ToObjectField*>({new TsObjectField(TsObjectFieldDescriptor(TsString(\"name\"), 0)new TsString(\"world!\"))})))}));\nconsole->getFieldValue(\"log\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({TsFunctionArg(\"sdjkl\", _foo->getFieldValue(\"bar\")->getFieldValue(\"str\"))}));\nconsole->getFieldValue(\"log\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({TsFunctionArg(\"sdjkl\", _foo->getFieldValue(\"baz\")->getFieldValue(\"name\"))}));\nauto _bar = new TsObject(1, TsCoreHelpers::toVector<ToObjectField*>({new TsObjectField(TsObjectFieldDescriptor(TsString(\"name\"), 0)new TsString(\"foo\")), new TsObjectField(TsObjectFieldDescriptor(TsString(\"sayName\"), 0)new TsFunction(\"\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nconsole->getFieldValue(\"log\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({TsFunctionArg(\"sdjkl\", _this->getFieldValue(\"getName\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({})))}));\nreturn NULL;\n})), new TsObjectField(TsObjectFieldDescriptor(TsString(\"getName\"), 0)new TsFunction(\"\",TsCoreHelpers::toVector<TsFunctionParam>({}), [=](TsObject* _this, std::vector<TsFunctionArg> args) -> TsObject* {\nreturn _this->getFieldValue(\"name\")\n;\n}))}));\n_bar->getFieldValue(\"sayName\")->invoke(TsCoreHelpers::toVector<TsFunctionArg>({}));\nreturn NULL;\n});\n_run->invoke(TsCoreHelpers::toVector<TsFunctionArg>({}));\n}",
}
